### ISSUE TITLE
inability to get spectrum working

### ISSUE THREAD

**Issue by cpherzog** [davidsd/sdpb#147]
Dear all,

I have been trying to get spectrum to work on the output from sdpb for a four current bootstrap.  

The command I use to call spectrum is

mpirun -n 32 $SDPBPATH/spectrum --input=$INPUT/sdpb.m --solution=$INPUT/output_out --threshold=1e-10 --format=PMP --output=$OUTPUT --precision=1024

The error I get is the following:

Error: Incompatible checkpoint file: '/cosma8/data/dp284/dc-herz2/chris/output_6576844_1/output_out/x_18.txt'.  Expected dimensions (0,1), but found (74,1)
--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 18 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------

I pass the input to sdpb using a Mathematica format file, and I believe sdpb is working correctly.  I'll copy the output of sdpb below.  I believe I am passing spectrum the correct locations for the sdpb input file and also the corresponding output files.  From the error message, it looks like the output files are not of the type expected.  Any suggestions would be appreciated.

Best,

Chris

Here is the output of sdpb

SDPB started at 2023-Nov-14 18:59:11
SDPB version: 2.5.1-119-gbac220b
SDP directory   : "/cosma8/data/dp284/dc-herz2/chris/output_6576844_1/output"
out directory   : "/cosma8/data/dp284/dc-herz2/chris/output_6576844_1/output_out"

Parameters:
maxIterations                = 1000
maxRuntime                   = 9223372036854775807
checkpointInterval           = 3600
findPrimalFeasible           = false
findDualFeasible             = false
detectPrimalFeasibleJump     = false
detectDualFeasibleJump       = false
precision(actual)            = 1024(1024)
dualityGapThreshold          = 1e-06
primalErrorThreshold         = 1e-30
dualErrorThreshold           = 1e-30
initialMatrixScalePrimal     = 1e+20
initialMatrixScaleDual       = 1e+20
feasibleCenteringParameter   = 0.1
infeasibleCenteringParameter = 0.3
stepLengthReduction          = 0.7
maxComplementarity           = 1e+100
initialCheckpointDir         = "/cosma8/data/dp284/dc-herz2/chris/output_6576844_1/output.ck"
checkpointDir                = "/cosma8/data/dp284/dc-herz2/chris/output_6576844_1/output.ck"
noFinalCheckpoint            = false
writeSolution                = x,y,X,Y
procsPerNode                 = 32
procGranularity              = 1
verbosity                    = 1

Performing a timing run, start at 2023-Nov-14 18:59:11

          time    mu     P-obj       D-obj      gap         P-err       p-err       D-err      P-step   D-step   beta
---------------------------------------------------------------------------------------------------------------------
1           17 1.0e+40  +157.       +157.       0.00       +1.00e+20   +2.83e+36   +9.22e+33   0.412    0.589    0.300
2           39 7.0e+39  +1.64e+25   -3.59e+16   1.00       +5.88e+19   +1.67e+36   +3.79e+33   0.434    0.551    0.300
3           61 4.9e+39  +2.35e+25   -8.08e+16   1.00       +3.32e+19   +9.43e+35   +1.70e+33   0.433    0.555    0.300
...
...
...
302       6632 5.3e-09  -7.63       -7.63       2.74e-06   +3.49e-250  +8.93e-196  +2.28e-52   0.497    0.497    0.100
303       6654 2.9e-09  -7.63       -7.63       1.51e-06   +3.35e-250  +2.55e-195  +1.15e-52   0.434    0.434    0.100
-----found primal-dual optimal solution-----

primalObjective = -7.632652807643978206020326212790134400248449090432881188443709035465242153253420961126130489798235080549300548942128184470450115934982828171707444603090036384459533088034648005274526095956970507734852624227120991483252482166242500795767158724801341936896393128440208123973309461019995909279499938335721098080349
dualObjective   = -7.632666896633324661111432240166672067601680886019595992108904987580512578934442340869720476034814353987899627584836715767717498610778381920417916047726421365219960538684002609785902261295699581887333751803617766988116162058694207408554514087722044367171699549252631954727894836959055778806183000611767797836191
dualityGap      = 9.229409943184742436043138138044463273373175900984579368347109768353061522061707070458663635543575825086883015289241479283087867318919798795785541178404279393370522608876433975796811834940563575877169298449443552986451481747737920410864767744516669966895070935417224957318550686720467241035394233360150219815023e-07
primalError     = 1.728657256361853088763612469769225373840185593490767933821671019294396239262028542046085232674471644839138765925668849651358759388537051356767072413540029977341309734124891964200549488377899844253590246772498719713329252738066563831178311438010843153940381190741293560830637395511581997540508606818138082852486e-193
dualError       = 6.505661272611547799779281014580091509038076329328321581568071368173478912094511941134524149094302940237301944027795068008952511680839791805958537842326452880640764831546419284830022349799732714357382422087203583294748572463006158157035159420958814756836254521478580514307173675776840378870714203741366055626973e-53

Saving checkpoint to    : "/cosma8/data/dp284/dc-herz2/chris/output_6576844_1/output.ck"
Saving solution to      : "/cosma8/data/dp284/dc-herz2/chris/output_6576844_1/output_out"

**Comment by davidsd** [comment]
Hi Chris, spectrum can only work when sdpb terminates with a primal-dual optimal solution. I noticed that your sdpb run ends with “max complementarity exceeded” instead.

**Comment by cpherzog** [comment]
Sorry, yes, let me try to reproduce the error with a primal dual optimal solution.  I don’t think that’s the issue though.

Best,

Chris

**Comment by cpherzog** [comment]
I have updated my description now with the correct sdpb output.  Sorry about the mistake earlier.  But yes, I am still having an issue, even with a primal dual optimal solution.  Chris

**Comment by vasdommes** [comment]
Hi Chris, thanks for reporting the bug!

I was able to reproduce this bug for [one of our sdpb tests](https://github.com/davidsd/sdpb/tree/master/test/data/end-to-end_tests/SingletScalar_cT_test_nmax6/primal_dual_optimal/json), will take a look at it soon.

**Comment by vasdommes** [comment]
Seems to be working with `mpirun -n 1`, @cpherzog could you please check that?

P.S. The same bug is in the initial 2.5.1 release too. Probably `spectrum` has never been working correctly in parallel (except for 1D case, where other ranks don't do anything)... @davidsd do you remember anything about it?

**Comment by cpherzog** [comment]
Thanks very much!  I no longer find that spectrum stops directly after startup.  

Probably I have some other problem now.  The output seems to be a list of “errors” of the form

{
    "zeros":
      [
      ],
    "error": "3892183.013028313046730698908504338966702894389784211799852912843990222167502431036866604570614891202246279104740100875545814191815976307136752796152468963882891743523888966189931348075637197786425612716117260755481895930806864251397815567885201215657575614095106681100376301420195704049766119103190726804139887"
  },

Best,

Chris

**Comment by cpherzog** [comment]
Martin Reehorst suggested the resolution to my missing zeros above.  It's that I set the threshold parameter in spectrum to be smaller than the primal-dual gap parameter in SDPB.  With the threshold parameter larger, spectrum succeeds in finding a number of zeros.

**Comment by vasdommes** [comment]
Closing this issue as answered.
- created separate issue for the initial problem #152 
- will add threshold parameter workaround to #143
- see also another known problem: #153
